connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -117,6 +118,20 @@ private[sql] class AvroDeserializer(
     val incompatibleMsg = errorPrefix +
         s"schema is incompatible (avroType = $avroType, sqlType = ${catalystType.sql})"
 
+    val realDataType = SchemaConverters.toSqlType(avroType).dataType
+    val confKey = SQLConf.LEGACY_AVRO_PREVENT_READING_INCORRECT_SAME_ENCODED_TYPES
+    val preventReadingIncorrectType = SQLConf.get.getConf(confKey)
+    def isNotExactType: Boolean = {
+      preventReadingIncorrectType && !realDataType.acceptsType(catalystType)
```

Comment:
It's weird to use `acceptsType` here, which is designed for expressions to define the expected type. I think we should just use `DataTypeUtils.sameType`.

Reply:
sounds good.

Reply:
just try to use DataTypeUtils.sameType, which causes regression and fails some tests. So maybe still using acceptsType is better/